src/syntaxtree.c

```diff
@@ -0,0 +1,2113 @@
+}
+
+Obj (* SyntaxTreeExprFuncs[256]) ( Expr expr );
+
+static inline Obj SyntaxTreeExpr(Expr expr)
+{
+    return (* SyntaxTreeExprFuncs[ TNUM_EXPR(expr) ])( expr );
+}
+
+Obj SyntaxTreeUnknownExpr(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("UnknownExpr",2);
+    AssPRec(result, RNamName("tnum"), INTOBJ_INT(TNUM_EXPR(expr)));
+    return result;
+}
+
+/* TODO: Find out why BoolExpr is special */
+Obj (* SyntaxTreeBoolExprFuncs[256]) ( Expr expr );
+
+Obj SyntaxTreeBoolExpr(Expr expr)
+{
+    return (* SyntaxTreeBoolExprFuncs[ TNUM_EXPR(expr) ])( expr );
+}
+
+Obj SyntaxTreeUnknownBool(Expr expr)
+{
+    /* compile the expression and check that the value is boolean          */
+    /* TODO: Check boolean? */
+    return SyntaxTreeExpr( expr );
+}
+
+extern Obj SyntaxTreeRefGVarFopy (Expr expr);
+
+Obj SyntaxTreeFunccall(Expr expr)
+{
+    Obj result;
+    Obj func;
+    Obj args, argi;
+    UInt narg, i;
+
+    result = NewSyntaxTreeNode("Funccall", 5);
+    /* TODO: If this is a gvar ref, put name? */
+    func = SyntaxTreeExpr( FUNC_CALL(expr) );
+
+    /*
+    if ( TNUM_EXPR( FUNC_CALL(expr) ) == T_REF_GVAR ) {
+    } */
+    AssPRec(result, RNamName("function"), func);
+
+    /* compile the argument expressions                                    */
+    narg = NARG_SIZE_CALL(SIZE_EXPR(expr));
+    args = NEW_PLIST(T_PLIST, narg);
+    SET_LEN_PLIST(args, narg);
+
+    for ( i = 1; i <= narg; i++ ) {
+        argi = SyntaxTreeExpr( ARGI_CALL( expr, i ) );
+        SET_ELM_PLIST(args, i, argi);
+        CHANGED_BAG(args);
+    }
+    AssPRec(result, RNamName("args"), args);
+    return result;
+}
+
+/* TODO: FuncCall options */
+Obj SyntaxTreeFunccallOpts(Expr expr)
+{
+    Obj result;
+    Obj opts;
+
+    result = NewSyntaxTreeNode("FunccallOpts", 2);
+    opts = SyntaxTreeExpr(ADDR_STAT(expr)[0]);
+    AssPRec(result, RNamName("opts"), opts);
+
+    return result;
+}
+
+static Obj SyntaxTreeFunc(Obj func);
+
+
+/* TODO: again? */
+Obj SyntaxTreeFuncExpr(Expr expr)
+{
+    Obj result;
+    Obj fexs;
+    Obj fexp;
+    Obj func;
+
+    result = NewSyntaxTreeNode("FuncExpr", 1);
+
+    fexs = FEXS_FUNC(CURR_FUNC);
+    fexp = ELM_PLIST(fexs, ((Int *)ADDR_EXPR(expr))[0]);
+    func = SyntaxTreeFunc(fexp);
+
+    AssPRec(result, RNamName("func"), func);
+
+    return result;
+}
+
+/* TODO: This is all the same, replace by SyntaxTreeBinaryOp(Expr,op) and
+ * SyntaxTreeUnary(Expr, op) */
+Obj SyntaxTreeOr(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("Or", 3);
+
+    AssPRec(result, RNamName("left"), SyntaxTreeBoolExpr( ADDR_EXPR(expr)[0]));
+    AssPRec(result, RNamName("right"), SyntaxTreeBoolExpr( ADDR_EXPR(expr)[1]));
+
+    return result;
+}
+
+Obj SyntaxTreeAnd(Expr expr)
+{
+    Obj result;
+
+    result = NewSyntaxTreeNode("And",3);
+
+    AssPRec(result, RNamName("left"), SyntaxTreeExpr(ADDR_EXPR(expr)[0]));
+    AssPRec(result, RNamName("right"), SyntaxTreeExpr(ADDR_EXPR(expr)[0]));
```

Comment:
yes, I was planning that anyway (there must be a comment around somewhere that says that these functions are all the same and they should be a macro or inline function)